File: tools/generate_description_tags.py
import os, sys, json, textwrap, re
import google.generativeai as genai

DEBUG = os.environ.get("DEBUG_GEMINI_GEN") == "1"

genai.configure(api_key=os.environ["GEMINI_API_KEY"])


def log_debug(msg: str):
    if DEBUG:
        print(f"[debug] {msg}", file=sys.stderr)


def build_prompt(payload: dict) -> str:
    # 入力: name, channel_desc, latest_video_title, latest_video_desc, video_tags(list)
    return textwrap.dedent(
        f"""
    あなたはVTuber/配信者の紹介文を作る編集者です。
    仕様:
    - 出力はJSONオブジェクト
    - description: 日本語で丁寧語、120字以内
    - tags: 配信ジャンル/特徴タグ 最大8件、ひらがな/カタカナ/漢字の単語にする（カンマ不要）
    - 個人情報・憶測は書かない。
    - 配信者 [配信者名]で検索を行い、その配信者の特徴を表すdescriptionとtagsを生成する。

    入力:
    配信者名: {payload.get("name","")}
    公式説明: {payload.get("channel_desc","")}
    """
    ).strip()


def generate(payload: dict):
    prompt = build_prompt(payload)
    log_debug(f"prompt:\n{prompt}")
    model = genai.GenerativeModel("gemini-2.5-flash")
    res = model.generate_content(prompt)

    # 生成結果全体をデバッグ出力
    try:
        log_debug(f"raw response: {res}")
        text = (res.text or "").strip()
    except Exception as e:
        log_debug(f"failed to read response text: {e}")
        text = ""

    log_debug(f"response text:\n{text}")

    # ```json ... ``` で囲まれている場合は中身だけを抜き出す
    if text.startswith("```"):
        # 正規表現の \s はそのまま。ダブルバックスラッシュだと文字列 "\s" にマッチしない。
        fenced = re.findall(r"```(?:json)?\s*(.*?)```", text, flags=re.S)
        if fenced:
            text = fenced[0].strip()
            log_debug(f"unfenced json:\n{text}")

    log_debug(f"抜き出し後のJSON:\n{text}")

    # 念のためJSONパース。失敗したらフォールバック。
    try:
        data = json.loads(text)
        desc = str(data.get("description", "")).strip().replace("\n", " ")
        tags = [t.strip() for t in data.get("tags") or [] if t.strip()]
    except Exception as e:
        log_debug(f"json parse error: {e}")
        desc = ""
        tags = []
    return {"description": desc, "tags": tags}


if __name__ == "__main__":
    # stdinからpayload(JSON)を受け取る
    payload = json.load(sys.stdin)
    result = generate(payload)
    json.dump(result, sys.stdout, ensure_ascii=False)

File: tools/generate-description-tags.ts
#!/usr/bin/env tsx
/**
 * VTuber/配信者の説明文とタグをAI生成するスクリプト
 * 標準入力(JSON) → 標準出力(JSON)
 */

import { OpenAI } from 'openai'

/* =====================
 * 設定
 * ===================== */
const DEBUG = process.env.DEBUG_GEMINI_GEN === '1'
const MODEL = 'gpt-5-mini'

// 外部レート制御が無い環境向けの安全弁
const RATE_LIMIT_WAIT_MS = Number(process.env.RATE_LIMIT_WAIT_MS ?? 2000)

/* =====================
 * Logger
 * ===================== */
const log = {
  info: (msg: string) => console.error(`[INFO] ${msg}`),
  error: (msg: string) => console.error(`[ERROR] ${msg}`),
  debug: (msg: string) => DEBUG && console.error(`[DEBUG] ${msg}`)
}

/* =====================
 * 型定義
 * ===================== */
interface Payload {
  name: string
  channel_desc: string
  latest_video_title?: string
  latest_video_desc?: string
  video_tags?: string[]
}

interface Result {
  description: string
  tags: string[]
}

/* =====================
 * OpenAI Client
 * ===================== */
const client = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY
})

/* =====================
 * Prompt
 * ===================== */
function buildPrompt(payload: Payload): string {
  return `
あなたはVTuber/配信者の「第三者紹介文」を作る編集者です。

制約:
- 本人視点の表現は禁止（「はじめまして」「私は」等NG）
- 出力は **JSONのみ**
- description:
  - 書き出しは必ず「この方は〇〇さんです。」
  - 日本語・丁寧語
  - 120字以内
- tags:
  - 最大8件
  - 単語のみ（ひらがな/カタカナ/漢字）
  - 必ず「VTuber」または「配信者」を含める
- 個人情報・憶測は書かない

入力:
配信者名: ${payload.name}
公式説明: ${payload.channel_desc}

出力フォーマット:
{
  "description": string,
  "tags": string[]
}
`.trim()
}

/* =====================
 * JSON Parse Helper
 * ===================== */
function safeParseJSON(text: string): Result {
  let jsonText = text.trim()

  // ```json``` フェンス除去（保険）
  if (jsonText.startsWith('```')) {
    const match = jsonText.match(/```(?:json)?\s*([\s\S]*?)```/)
    if (match) jsonText = match[1].trim()
  }

  try {
    const data = JSON.parse(jsonText)
    return {
      description: String(data.description || '').replace(/\n/g, ' ').trim(),
      tags: Array.isArray(data.tags)
        ? data.tags.map((t: string) => t.trim()).filter(Boolean)
        : []
    }
  } catch (e) {
    log.error(`JSON parse failed`)
    log.debug(jsonText)
    return { description: '', tags: [] }
  }
}

/* =====================
 * Generate
 * ===================== */
async function generate(payload: Payload): Promise<Result> {
  const prompt = buildPrompt(payload)
  log.debug(`Prompt:\n${prompt}`)

  try {
    const res = await client.responses.create({
      model: MODEL,
      input: [
        { role: 'system', content: 'You are a helpful assistant.' },
        { role: 'user', content: prompt }
      ]
    })

    const output = res.output_text?.trim() ?? ''
    log.info(`AI response received (${output.length} chars)`)

    return safeParseJSON(output)
  } catch (e) {
    log.error(`OpenAI API error: ${e}`)
    return { description: '', tags: [] }
  }
}

/* =====================
 * Main
 * ===================== */
async function main() {
  try {
    log.info('Reading stdin...')
    let input = ''
    process.stdin.setEncoding('utf8')
    for await (const chunk of process.stdin) input += chunk

    const payload: Payload = JSON.parse(input)
    log.info(`Payload loaded: ${payload.name}`)

    // レート制御（外部で制御できない環境向けの安全弁）
    log.info(`Waiting ${RATE_LIMIT_WAIT_MS}ms for rate limit control...`)
    await new Promise(resolve => setTimeout(resolve, RATE_LIMIT_WAIT_MS))

    const result = await generate(payload)

    // 常にJSONを返す（失敗時も）
    console.log(JSON.stringify(result))
  } catch (e) {
    log.error(`Fatal error: ${e}`)
    console.log(JSON.stringify({ description: '', tags: [] }))
    process.exit(1)
  }
}

main()
